### ISSUE TITLE
Absolute paths + local server vs. relative paths + no server + workarounds

### ISSUE THREAD

**Issue by gvwilson** [gvwilson/new-lesson-example#28]
@gvwilson and @wking will use this issue to build the pro and con of two different ways to handle local builds.

**Comment by wking** [comment]
Here's my current understanding:

| | absolute links + site.path config | absolute links with fixed site.path | relative links | relative links with subdir layout |
| --- | --- | --- | --- | --- |
| Local server + any browser | + | - | + | + |
| Non-Firefox | - | - | + | + |
| Stock Firefox | - | - | - | + |
| Firefox with disabled file URI security | - | - | + | + |

Where:

* `site.path` config is "tweak `site.path` for dev, e.g. with a `_config_dev.yml`.
* a subdir layout satisfies [this][1].

So relative links give more flexibility (working under more conditions) than absolute links, and for maximum flexibility we should follow Firefox's file URI security.

Cons for relative links are "how ugly is the support?"  Ignoring Firefox's file URI security gives the current #27, which feels like an implementation wash with the current absolute links.  I haven't implemented a layout compatible with the file-URI restriction, so I'm not sure how clean that will be yet.

[1]: https://developer.mozilla.org/en-US/docs/Same-origin_policy_for_file:_URIs

**Comment by gdevenyi** [comment]
Is there a toy example for each of the "options" in the table? I think I've run into these differences before hosting our labgroup site, but I'm not sure...

**Comment by qjcg** [comment]
I'd vote for the following in descending order of preference:

1. relative paths with subdir layout
2. absolute paths + local server

**Comment by wking** [comment]
On Thu, May 26, 2016 at 06:33:59AM -0700, Gabriel A. Devenyi wrote:
> Is there a toy example for each of the "options" in the table? I
> think I've run into these differences before hosting our labgroup
> site, but I'm not sure...

The current version of this repo is “absolute links + site.path
config”.  “absolute links with fixed site.path” would be this repo
with hard-coded URLs instead of [1].  “relative links” is #27.
“relative links with subdir layout” has not been written yet.

[1]: https://github.com/gvwilson/new-lesson-example/blob/gh-pages/_config.yml#L11-L12

**Comment by wking** [comment]
On Fri, May 27, 2016 at 11:52:22AM -0700, John Gosset wrote:
> I'd vote for the following in descending order of preference:
> 
> 1. relative paths with subdir layout
> 2. absolute paths + local server

Can you explain why you like absolute paths (e.g. current state)
better than relative links (e.g. #27)?  The only situation in which
#27 doesn't work is “stock Firefox + file://”, and those folks can do
“relative paths + local server” until someone figures out “relative
paths with a subdir layout”.

**Comment by qjcg** [comment]
Hi @wking : 

> Can you explain why you like absolute paths (e.g. current state) better than relative links (e.g. #27)? The only situation in which #27 doesn't work is “stock Firefox + file://”, and those folks can do “relative links + local server”

I disliked the idea of suggesting people mess around with security settings in about:config. If we'll be suggesting they do "relative links + local server" instead, that's better. I edited my vote/comment above to reflect this.

> until someone figures out “relative paths with a subdir layout”

Is there some hidden complexity with this? My reading of the [firefox file:// same-origin policy](https://developer.mozilla.org/en-US/docs/Same-origin_policy_for_file:_URIs) is that we'd simply need to store all of our assets in the same directory or a subdir to satisfy the policy.

**Comment by wking** [comment]
On Fri, May 27, 2016 at 12:18:08PM -0700, John Gosset wrote:
> > until someone figures out “relative paths with a subdir layout”
>
> Is there some hidden complexity with this? My reading of the firefox
> file:// same-origin policy is that we'd simply need to store all of
> our assets in the same directory or a subdir to satisfy the policy.

Yeah, I'm just not particularly excited about moving assets/fonts to
assets/css/fonts (or whatever it would take).  And I haven't had time
to figure out a more palatable solution.

**Comment by qjcg** [comment]
> I'm just not particularly excited about moving assets/fonts to assets/css/fonts (or whatever it would take).  And I haven't had time to figure out a more palatable solution.

Fair enough. This being something of a [DoOcracy](https://communitywiki.org/DoOcracy), I'll leave it at that for now unless I can find some time to look more closely and submit a proof-of-concept PR.
